File: src/main/java/uk/gov/justice/digital/hmpps/riskprofiler/exception/PrisonException.java
package uk.gov.justice.digital.hmpps.riskprofiler.exception;

import java.util.function.Supplier;

public class PrisonException extends RuntimeException implements Supplier<PrisonException> {
    private static final String INVALID_PRISON = "Prison [%s] is invalid.";
    private static final String PRISON_EXISTS = "Prison [%s] is already present.";

    public static PrisonException exists(String id) {
        return new PrisonException(String.format(PRISON_EXISTS, id));
    }

    public static PrisonException withId(String id) {
        return new PrisonException(String.format(INVALID_PRISON, id));
    }

    public static PrisonException withMessage(String message) {
        return new PrisonException(message);
    }

    public static PrisonException withMessage(String message, Object... args) {
        return new PrisonException(String.format(message, args));
    }

    public PrisonException(String message) {
        super(message);
    }

    @Override
    public PrisonException get() {
        return new PrisonException(getMessage());
    }

    @Override
    public synchronized Throwable fillInStackTrace() {
        return this;
    }
}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/riskprofiler/exception/PrisonException.kt
package uk.gov.justice.digital.hmpps.riskprofiler.exception

import java.util.function.Supplier

class PrisonException(message: String?) : RuntimeException(message), Supplier<PrisonException> {
  override fun get(): PrisonException {
    return PrisonException(message)
  }

  @Synchronized
  override fun fillInStackTrace(): Throwable {
    return this
  }

  companion object {
    private const val INVALID_PRISON = "Prison [%s] is invalid."
    private const val PRISON_EXISTS = "Prison [%s] is already present."

    @JvmStatic
    fun exists(id: String?): PrisonException {
      return PrisonException(String.format(PRISON_EXISTS, id))
    }

    @JvmStatic
    fun withId(id: String?): PrisonException {
      return PrisonException(String.format(INVALID_PRISON, id))
    }

    fun withMessage(message: String?): PrisonException {
      return PrisonException(message)
    }

    fun withMessage(message: String?, vararg args: Any?): PrisonException {
      return PrisonException(String.format(message!!, *args))
    }
  }
}
